src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -621,17 +619,16 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 return;
             }
 
-            let mut trait_type = trait_ref.self_ty();
+            let mut suggested_ty = trait_ref.self_ty();
 
             for refs_remaining in 0..refs_number {
-                if let ty::Ref(_, t_type, _) = trait_type.kind {
-                    trait_type = t_type;
+                if let ty::Ref(_, inner_ty, _) = suggested_ty.kind {
+                    suggested_ty = inner_ty;
 
-                    let new_obligation = self.mk_obligation_for_def_id(
-                        trait_ref.def_id,
-                        trait_type,
-                        ObligationCause::dummy(),
+                    let new_obligation = self.mk_trait_obligation_with_new_self_ty(
                         obligation.param_env,
+                        trait_ref,
+                        suggested_ty,
```

Comment:
ignoring the previous obligation entirely? We used to use the same cause, right?

Is this what causes the "note: `std::convert::From<std::string::String>` is implemented for `&mut str`, but not for `&str`" to go away?

Reply:
No. That comes from propagating substitutions correctly. `From<String>` is not implemented for any of `&mut str` or `&str`.

The cause for the existing obligation is where it originates in the source code. There is no underlying cause for the new obligation we create, since we're just trying various "self" types for diagnostic hints. From this POV, obligations that are only used speculatively should have no cause code.